### PR TITLE
Try to bring back syscall filter and StreamHandler

### PR DESCRIPTION
**Pull request by ucgJhe**
as title. other OS API filter will be next

### CODE REVIEW

File: qiling/core.py [hunk 1]
```diff
@@ -133,9 +133,11 @@ def __init__(
         if self.log_dir != "" and type(self.log_dir) == str:
             _logger = ql_setup_logging_env(self)    
             self.log_file_fd = _logger
-        else:
-            self.log_dir = None     
-        
+        elif self.console == True:
```
Comment (xwings) [inline qiling/core.py:136]: this only work during startup? what if we want to turn on or off console during execution or within a hook?

File: qiling/core_utils.py [hunk 2]
```diff
@@ -46,9 +44,9 @@ def nprint(self, *args, **kw):
 
             # setup filter for logger
             # FIXME: only works for logging due to we might need runtime disable nprint, it should be a global filter not only syscall
-            # if self.strace != None and self.output == QL_OUTPUT.DEFAULT:
-            #     from .utils import Ql_filter
-            #     self.log_file_fd.addFilter(ql_filter(self.filter))
+            if self.filter != None and self.output == QL_OUTPUT.DEFAULT:
+                from .utils import Ql_filter
+                self.log_file_fd.addFilter(Ql_filter(self.filter))
```
Comment (xwings) [inline qiling/core_utils.py:49]: QL filter is a bad name, i admin i choose the name but we need a better one

File: qiling/core.py (hunk 1 above)
Comment (ucgJhe) [reply qiling/core.py:136]: add a simple console checking will let you do it, turn off or on console logging by `ql.console = True` or `ql.console = False` , is this enough or we need more granular ?

Comment (ucgJhe) [reply qiling/core.py:136]: like this
```python
import sys
sys.path.append("..")
from qiling.os.posix.syscall.fcntl import ql_syscall_openat
from qiling import *

def my_openat(ql, *args):
    ql.console = False
    ql_syscall_openat(ql, *args)
    ql.console = True


if __name__ == "__main__":
    ql = Qiling(["examples/rootfs/x8664_linux/bin/x8664_hello"], "examples/rootfs/x8664_linux", console=True)
    ql.set_syscall("openat", my_openat)
    ql.run()

```

Comment (xwings) [reply qiling/core.py:136]: can you test it this way ?
```
def my_openat(ql, *args):
    ql.console = False
    ql,nprint("Shows nothing") 
    ql.console = True
    ql,nprint("Shows something")
```

Comment (ucgJhe) [reply qiling/core.py:136]: image

Comment (xwings) [reply qiling/core.py:136]: Nice

File: qiling/core_utils.py (hunk 2 above)
Comment (xwings) [reply qiling/core_utils.py:49]: Now we need to figure this out. maybe just call it filter

Comment (ucgJhe) [reply qiling/core_utils.py]: how about that we made Ql_filter as a setup function for now, maybe the name will be change after other filters is implemented in the future

File: qiling/core.py [hunk 3]
```diff
@@ -133,9 +133,11 @@ def __init__(
...
+            _logger = ql_setup_logging_stream(self)
+            self.log_file_fd = _logger
+            # self.log_dir = None
```
Comment (xwings) [inline qiling/core.py:139]: this need to be  out.

File: qiling/core_utils.py [hunk 4]
```diff
@@ -46,13 +44,14 @@ def nprint(self, *args, **kw):
...
-            #     self.log_file_fd.addFilter(ql_filter(self.filter))
+            if self.filter != None and self.output == QL_OUTPUT.DEFAULT:
+                from .utils import ql_setup_filter
+                self.log_file_fd.addFilter(ql_setup_filter(self.filter))
 
             msg = args[0]
             msg += kw["end"] if kw.get("end", None) != None else os.linesep
-            fd.info(msg)
+            if self.console == True:
+                fd.info(msg)
```
Comment (xwings) [inline qiling/core_utils.py:54]: One more question. 

if ql.console == False will it still write to the log ?

File: qiling/core.py (hunk 3 above)
Comment (ucgJhe) [reply qiling/core.py:139]: you mean the commented part or whole section ?

Comment (xwings) [reply qiling/core.py:139]: yes comment one. unless there is a specific reason. if year please write it down

File: qiling/core_utils.py (hunk 4 above)
Comment (xwings) [reply qiling/core_utils.py:54]: I guess this is the last reqeust

Once we confirm this we are good to go.

Comment (ucgJhe) [reply qiling/core_utils.py]: done it by add a FalseFilter for all stream handler when console=False, at the same time still write to disk logging file

Comment (ucgJhe) [reply qiling/core_utils.py]: testing code:
```
#!/usr/bin/env python3
# 
# Cross Platform and Multi Architecture Advanced Binary Emulation Framework
# Built on top of Unicorn emulator (www.unicorn-engine.org) 
import sys
sys.path.append("..")
from qiling.os.posix.syscall.fcntl import ql_syscall_openat
from qiling import *

def my_openat(ql, *args):
    ql.console = True
    ql.nprint("Shows something")
    ql.console = False
    ql.nprint("Shows nothing") 
    ql_syscall_openat(ql, *args)

if __name__ == "__main__":
    ql = Qiling(["examples/rootfs/x8664_linux/bin/x8664_hello"], "examples/rootfs/x8664_linux", console=False, log_dir='qlog')
    ql.set_syscall("openat", my_openat)
    ql.run()
```

Comment (ucgJhe) [reply qiling/core_utils.py]: result from console:
image

Comment (ucgJhe) [reply qiling/core_utils.py]: result from log:
image

**Comment by xwings** [comment]
Nice, thanks for the work.

File: qiling/utils.py [hunk 5]
```diff
@@ -194,4 +209,21 @@ def ql_setup_logging_file(ql_mode, log_file_path, logger=None):
...
+            super().__init__()
+            # accept list or string func_names so you can use it in qltool and programming
+            self.filter_list = func_names.strip().split(",") if isinstance(func_names, str) else func_names
+
+        def filter(self, record):
+            return any((record.getMessage().startswith(each) for each in self.filter_list))
+
+    return _filter(func_names)
+
+class _FalseFilter(logging.Filter):
```
Comment (xwings) [inline qiling/utils.py:225]: the naming is not consistent. can we have a better naming ?

Comment (ucgJhe) [reply qiling/utils.py:225]: how about merge it to ql_setup_filter ?

Comment (xwings) [reply qiling/utils.py:225]: i guess that is a good idea

**Comment by xwings** [comment]
Thanks man. Lets merge

**Pull request merged**